// @bacons/xcode/build/api/PBXFileSystemSynchronizedRootGroup.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PBXFileSystemSynchronizedRootGroup = void 0;
const json = __importStar(require("../json/types"));
const AbstractObject_1 = require("./AbstractObject");
class PBXFileSystemSynchronizedRootGroup extends AbstractObject_1.AbstractObject {
    static is(object) {
        return object.isa === PBXFileSystemSynchronizedRootGroup.isa;
    }
    static create(project, opts) {
        return project.createModel({
            isa: PBXFileSystemSynchronizedRootGroup.isa,
            ...opts,
        });
    }
    getObjectProps() {
        return {
            exceptions: [String],
        };
    }
    removeFromProject() {
        // Remove all exceptions that are only referenced by this group
        if (this.props.exceptions) {
            for (const exception of [...this.props.exceptions]) {
                const referrers = exception.getReferrers();
                if (referrers.length === 1 && referrers[0].uuid === this.uuid) {
                    exception.removeFromProject();
                }
            }
        }
        return super.removeFromProject();
    }
}
exports.PBXFileSystemSynchronizedRootGroup = PBXFileSystemSynchronizedRootGroup;
PBXFileSystemSynchronizedRootGroup.isa = json.ISA.PBXFileSystemSynchronizedRootGroup;
