// react-native-device-activity/build/ReactNativeDeviceActivity.types.js
export const AuthorizationStatus = {
    /** Authorization status has not been determined yet */
    notDetermined: 0,
    /** Authorization has been denied */
    denied: 1,
    /** Authorization has been approved */
    approved: 2,
};
export const UIBlurEffectStyle = {
    extraLight: 0,
    light: 1,
    dark: 2,
    // @available(iOS 10.0, *)
    regular: 4,
    // @available(iOS 10.0, *)
    prominent: 5,
    // @available(iOS 13.0, *)
    systemUltraThinMaterial: 6,
    // @available(iOS 13.0, *)
    systemThinMaterial: 7,
    // @available(iOS 13.0, *)
    systemMaterial: 8,
    // @available(iOS 13.0, *)
    systemThickMaterial: 9,
    // @available(iOS 13.0, *)
    systemChromeMaterial: 10,
    // @available(iOS 13.0, *)
    systemUltraThinMaterialLight: 11,
    // @available(iOS 13.0, *)
    systemThinMaterialLight: 12,
    // @available(iOS 13.0, *)
    systemMaterialLight: 13,
    // @available(iOS 13.0, *)
    systemThickMaterialLight: 14,
    // @available(iOS 13.0, *)
    systemChromeMaterialLight: 15,
    // @available(iOS 13.0, *)
    systemUltraThinMaterialDark: 16,
    // @available(iOS 13.0, *)
    systemThinMaterialDark: 17,
    // @available(iOS 13.0, *)
    systemMaterialDark: 18,
    // @available(iOS 13.0, *)
    systemThickMaterialDark: 19,
    // @available(iOS 13.0, *)
    systemChromeMaterialDark: 20,
};
